services/download/src/main/java/app/coronawarn/server/services/download/FederationBatchProcessor.java

```diff
@@ -0,0 +1,146 @@
+   * @param diagnosisKeyService     A {@link DiagnosisKeyService} for storing retrieved diagnosis keys.
+   * @param federationGatewayClient A {@link FederationGatewayClient} for retrieving federation diagnosis key batches.
+   */
+  public FederationBatchProcessor(FederationBatchInfoService batchInfoService,
+      DiagnosisKeyService diagnosisKeyService, FederationGatewayClient federationGatewayClient) {
+    this.batchInfoService = batchInfoService;
+    this.diagnosisKeyService = diagnosisKeyService;
+    this.federationGatewayClient = federationGatewayClient;
+  }
+
+  /**
+   * Stores the batch info for the specified date. Its status is set to {@link FederationBatchStatus#UNPROCESSED}.
+   *
+   * @param date The date for which the first batch info is stored.
+   */
+  public void saveFirstBatchInfoForDate(LocalDate date) {
+    try {
+      logger.info("Downloading first batch for date {}", date);
+      BatchDownloadResponse response = federationGatewayClient.getDiagnosisKeys(date.format(ISO_LOCAL_DATE));
+      batchInfoService.save(new FederationBatchInfo(response.getBatchTag(), date));
+    } catch (Exception e) {
+      logger.error("Downloading batch for date {} failed", date, e);
+    }
+  }
+
+  /**
+   * Downloads and processes all batches from the federation gateway that have previously been marked with the status
+   * value {@link FederationBatchStatus#ERROR}.
+   */
+  public void processErrorFederationBatches() {
+    List<FederationBatchInfo> federationBatchInfosWithError = batchInfoService.findByStatus(ERROR);
+    federationBatchInfosWithError.forEach(this::retryProcessingBatch);
+  }
+
+  private void retryProcessingBatch(FederationBatchInfo federationBatchInfo) {
+    try {
+      processBatchAndReturnNextBatchId(federationBatchInfo, ERROR_WONT_RETRY)
+          .ifPresent(nextBatchTag ->
+              batchInfoService.save(new FederationBatchInfo(nextBatchTag, federationBatchInfo.getDate())));
+    } catch (Exception e) {
+      logger.error("Failed to save next federation batch info for processing. Will not try again.", e);
+      batchInfoService.updateStatus(federationBatchInfo, ERROR_WONT_RETRY);
+    }
+  }
+
+  /**
+   * Downloads and processes all batches from the federation gateway that have previously been marked with status value
+   * {@link FederationBatchStatus#UNPROCESSED}.
+   */
+  public void processUnprocessedFederationBatches() {
+    Deque<FederationBatchInfo> unprocessedBatches = new LinkedList<>(batchInfoService.findByStatus(UNPROCESSED));
+
+    while (!unprocessedBatches.isEmpty()) {
+      FederationBatchInfo currentBatch = unprocessedBatches.remove();
+      processBatchAndReturnNextBatchId(currentBatch, ERROR)
+          .ifPresent(nextBatchTag ->
+              unprocessedBatches.add(new FederationBatchInfo(nextBatchTag, currentBatch.getDate())));
+    }
+  }
+
+  private Optional<String> processBatchAndReturnNextBatchId(
```

Comment:
I find a bit confusing handling the update exception inside this method. Specially when one of the callers (`retryProcessingBatch`) also catches the error and does the error status update. Maybe for more clarity would be good to handle the exception only on the callers (`retryProcessingBatch` and `processUnprocessedFederationBatches`) and simply throw the exception here. 

If we need different log messages (or even different error handling per type of error) we should then use different Exception classes. So other than having a `try` block with a generic `Exception` catch, we could do something like:

```java
try {
  processBatchAndReturnNextBatchId(federationBatchInfo)
    // do something with the result
} catch (EFGSRequestException e)
  // ...
} catch (DBException e) {
  // ... 
}
```

This is more a refactoring suggestion rather than an issue. Based on what I understood from this part of the code I don't see any errors that could arise with the current implementation.